python/paddle/fluid/framework.py

```diff
@@ -1319,25 +1319,28 @@ def set_op_role_var(self, var_name):
         self._op_role_var = [var_name]
 
     @contextlib.contextmanager
-    def optimized_guard(self, var):
+    def optimized_guard(self, param_and_grads):
         """
         A with guard to set :code:`Optimization` :code:`OpRole` and
         :code:`OpRoleVar` automatically.
 
         Notes: This is a very low level API. Users should not use it directly.
 
         Args:
-            var(Variable|str): The variable (name) to be optimized.
+            param_and_grads(list): The variables (names) to be optimized.
 
         Examples:
 
             >>> p, g = backward(...)
-            >>> with program.optimized_guard(p):
+            >>> with program.optimized_guard([p,g]):
             >>>     p = p - 0.001 * g
         """
         OpRole = core.op_proto_and_checker_maker.OpRole
         self._current_role = OpRole.Optimize
-        self._op_role_var = [var.name if isinstance(var, Variable) else var]
+        self._op_role_var = [
```

Comment:
Why we need to store parameters and gradients?

Reply:
<img width="870" alt="42629173-1d2bf772-8605-11e8-8d0b-f20980cf73e5" src="https://user-images.githubusercontent.com/30176695/42672908-4cfb539e-869b-11e8-83dc-f2e8607e0021.png">

In this case, `fc_0.b_0`'s gradience name has been changed, so if we still use gradient`fc_0.b_0@GRAD` to decide this `sgd` in which device, we will encounter errors.
So, in this PR, I use `grad` but not `GradVarName(params[0])` to get it's belong device.
https://github.com/PaddlePaddle/Paddle/pull/11406/files/7cf836f4ba4353d8ba4247ca09e904098a43edf5#diff-06c27dc69562c2f50b53409969b0a9b5R435

Reply:
I see, cool